crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3229,3 +3230,155 @@ mod tests {
+                10 => Ty::BytesLiteral(""),
+                11 => Ty::BytesLiteral("a"),
+                12 => Ty::BytesLiteral("aa"),
+                13 => Ty::BytesLiteral("\x00"),
+                14 => Ty::BuiltinInstance("str"),
+                15 => Ty::BuiltinInstance("int"),
+                16 => Ty::BuiltinInstance("bool"),
+                17 => Ty::BuiltinInstance("object"),
+                18 => Ty::BuiltinClassLiteral("str"),
+                19 => Ty::BuiltinClassLiteral("int"),
+                20 => Ty::BuiltinClassLiteral("bool"),
+                21 => Ty::BuiltinClassLiteral("object"),
+                22 => Ty::Union(vec![
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                ]),
+                23 => Ty::Union(vec![
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                ]),
+                24 => Ty::Union(vec![Ty::BuiltinInstance("int"), Ty::None]),
+                25 => Ty::Union(vec![Ty::IntLiteral(1), Ty::BuiltinInstance("str")]),
+                26 => Ty::Intersection {
+                    pos: vec![Ty::BuiltinInstance("int")],
+                    neg: vec![Ty::IntLiteral(i64::arbitrary(g))],
+                },
+                27 => Ty::Intersection {
+                    pos: vec![Ty::BuiltinInstance("str")],
+                    neg: vec![Ty::StringLiteral("a")],
+                },
+                28 => Ty::Tuple(vec![]),
+                29 => Ty::Tuple(vec![Ty::IntLiteral(i64::arbitrary(g))]),
+                30 => Ty::Tuple(vec![
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                    Ty::IntLiteral(i64::arbitrary(g)),
+                ]),
+                31 => Ty::Tuple(vec![Ty::BuiltinInstance("str"), Ty::BuiltinInstance("str")]),
+                32 => Ty::Tuple(vec![Ty::BuiltinInstance("str"), Ty::BuiltinInstance("int")]),
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    #[quickcheck]
+    fn equivalent_to_is_reflexive(t: Ty) -> bool {
+        let db = setup_db();
+        let t = t.into_type(&db);
+
+        t.is_equivalent_to(&db, t)
+    }
+
+    #[quickcheck]
+    fn subtype_of_is_reflexive(t: Ty) -> bool {
+        let db = setup_db();
+        let t = t.into_type(&db);
+
+        t.is_subtype_of(&db, t)
+    }
+
+    #[quickcheck]
+    fn subtype_of_is_antisymmetric(t1: Ty, t2: Ty) -> bool {
+        let db = setup_db();
+
+        let t1 = t1.into_type(&db);
+        let t2 = t2.into_type(&db);
+
+        !(t1.is_subtype_of(&db, t2) && t2.is_subtype_of(&db, t1)) || (t1 == t2)
+    }
+
+    #[quickcheck]
+    fn subtype_of_is_transitive(t1: Ty, t2: Ty, t3: Ty) -> bool {
+        let db = setup_db();
+
+        let t1 = t1.into_type(&db);
+        let t2 = t2.into_type(&db);
+        let t3 = t3.into_type(&db);
```

Comment:
If we want to move forward with this, we could probably auto-generate this boilerplate in each test using a macro.